image/image_test.go

```diff
@@ -3457,3 +3457,254 @@ func (s *imageSuite) TestSetupSeedCore20DelegatedSnap(c *C) {
+	seedsnapsdir := filepath.Join(seeddir, "snaps")
+	essSnaps, runSnaps, _ := s.loadSeed(c, seeddir)
+	c.Check(essSnaps, HasLen, 3)
+	c.Check(runSnaps, HasLen, 1)
+
+	// check the files are in place
+	for i, name := range []string{"core_x1.snap", "pc-kernel", "pc"} {
+		channel := stableChannel
+		info := s.AssertedSnapInfo(name)
+		var pinfo snap.PlaceInfo = info
+		var sideInfo *snap.SideInfo
+		var snapType snap.Type
+		if info == nil {
+			switch name {
+			case "core_x1.snap":
+				pinfo = snap.MinimalPlaceInfo("core", snap.R(-1))
+				sideInfo = &snap.SideInfo{
+					RealName: "core",
+				}
+				channel = ""
+				snapType = snap.TypeOS
+			}
+		} else {
+			sideInfo = &info.SideInfo
+			snapType = info.Type()
+		}
+
+		fn := pinfo.Filename()
+		p := filepath.Join(seedsnapsdir, fn)
+		c.Check(p, testutil.FilePresent)
+		c.Check(essSnaps[i], DeepEquals, &seed.Snap{
+			Path: p,
+
+			SideInfo: sideInfo,
+
+			EssentialType: snapType,
+			Essential:     true,
+			Required:      true,
+
+			Channel: channel,
+		})
+	}
+	c.Check(runSnaps[0], DeepEquals, &seed.Snap{
+		Path: filepath.Join(seedsnapsdir, "required-snap1_x1.snap"),
+
+		SideInfo: &snap.SideInfo{
+			RealName: "required-snap1",
+		},
+		Required: true,
+	})
+	c.Check(runSnaps[0].Path, testutil.FilePresent)
+
+	l, err := ioutil.ReadDir(seedsnapsdir)
+	c.Assert(err, IsNil)
+	c.Check(l, HasLen, 4)
+
+	// check the downloads
+	c.Check(s.storeActionsBunchSizes, DeepEquals, []int{2})
+	c.Check(s.storeActions[0], DeepEquals, &store.SnapAction{
+		Action:       "download",
+		InstanceName: "pc-kernel",
+		Revision:     revisions["pc-kernel"],
+		Flags:        store.SnapActionIgnoreValidation,
+	})
+	c.Check(s.storeActions[1], DeepEquals, &store.SnapAction{
+		Action:       "download",
+		InstanceName: "pc",
+		Revision:     revisions["pc"],
+		Flags:        store.SnapActionIgnoreValidation,
+	})
+	return nil
+}
+
+func (s *imageSuite) TestSetupSeedSnapRevisionsWithLocalSnapFails(c *C) {
+	// It doesn't make sense to use a local snap when doing a reproducible build,
+	// so if a revision is provided, and we are trying to provide that snap locally,
+	// then we should return an error.
+	// Our helper creates two local snaps
+	// 1. core.
+	// 2. required-snap.
+	// So lets provide a revision for one of them and it should then fail
+	err := s.testSetupSeedWithMixedSnapsAndRevisions(c, map[string]snap.Revision{
+		"pc-kernel": snap.R(1),
+		"pc":        snap.R(13),
+		"core":      snap.R(5),
+	})
+	c.Check(err, ErrorMatches, `cannot use snap .*/snapsrc/core_16.04_all.snap for image, unknown/local revision does not match the value specified by revisions file \(unset != 5\)`)
+}
+
+func (s *imageSuite) TestSetupSeedSnapRevisionsWithLocalSnapHappy(c *C) {
+	// Make sure we can still provide specific revisions for snaps that are
+	// non-local.
+	err := s.testSetupSeedWithMixedSnapsAndRevisions(c, map[string]snap.Revision{
+		"pc-kernel": snap.R(15),
+		"pc":        snap.R(28),
```

Comment:
should this be the test with core set to x1?

Reply:
do we need a test with a local file that corresponds to a store revision as well? we should have  a test like that already I hope that we could derive this kind of test from, I think TestSetupSeedLocalSnapsWithStoreAsserts might be that test

Reply:
Yes indeed, we definitely need another unit test to verify that case, i'll see how much I can reuse from the mentioned test